### PR TITLE
Fast Evolutionary Algorithm Support

### PR DESCRIPTION
**Pull request by DavidKoplow**
A more stable and much faster implementation of evolution strategies than was presented in https://arxiv.org/abs/2509.24372.

Features:

* ES strategies support for OpenRLHF
* 10-30x faster than the original implementation.
* Uses reversible floating point addition/subtraction by upcasting to fp32 prior to mutations to enable constant memory overhead.
* Decentralized learning that can scale to many clusters
* Zero slowdown during vLLM inference engine through direct parameter mutations.
* Very fast model mutation within workers, and support for multiple concurrent mutations across workers.
* Optionally add the current model's outputs to the ES update computation to improve training stability

### CODE REVIEW

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+            optimizer.step()
+            p.grad = None
+
+            # Move optimizer state back to CPU to save memory
+            st = optimizer.state.get(p)
+            if st:
+                for k, v in list(st.items()):
+                    if isinstance(v, torch.Tensor):
+                        st[k] = v.to("cpu")
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:161]: This is actually much faster since it allows higher throughput for vLLM and only optimizer states are moved to CPU, which doesn't cause any slowdown for standard ES since it uses SGD and thus does not need to save any gradient estimates between steps.

File: examples/python/es_countdown_heuristic.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    def __call__(self, queries: List[str], prompts: List[str], labels: List[str]) -> Dict[str, List[float]]:
+        rewards = []
+        for query, label in zip(queries, labels):
+            try:
+                # 1. Safely handle pre-parsed lists from the dataset
+                parsed = ast.literal_eval(label) if isinstance(label, str) else label
+
+                expected_nums = sorted(int(n) for n in parsed[0])
+                tgt = int(parsed[1][0]) if isinstance(parsed[1], list) else int(parsed[1])
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply examples/python/es_countdown_heuristic.py:21]: This doesn't throw an error because of the example dataset. The proposed fix would be incorrect; it would indicate an issue with the dataset.

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+from .vllm_worker_wrap import WorkerWrap
+
+STABILIZE_SEED = -1
+
+
+def _noise(p: torch.Tensor, name: str, seed: int) -> torch.Tensor:
+
+    g = torch.Generator(device=p.device)
+    h = int.from_bytes(hashlib.sha256(name.encode()).digest()[:4], "little")
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:26]: This takes a couple of microseconds, and there are generally under a thousand parameters in a model so this isn't worth caching.

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    g.manual_seed((seed ^ h) & 0xFFFFFFFF)
+    n = torch.normal(0, 1, size=p.shape, device=p.device, generator=g, dtype=p.dtype)
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py]: The noise needs to be the same type as the original datatype (assuming fp16 or bf16) in order for the computation to be reversible in fp32. There is some precision loss, but it doesn't matter so long as the computation is reversible, so the model can run the next seed without having its weights get out of sync with the other models on the cluster.

**Comment by DavidKoplow** [comment]
@gemini-code-assist re-run the review

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                    g.add_(_noise(p.data, name, seed).float(), alpha=float(w))
+            g.div_(max(1, len(updates)))
+
+            # ES uses negative gradient (we want to move in direction of positive scores)
+            p.grad = (-g).to(p.dtype)
+
+            if clip > 0:
+                torch.nn.utils.clip_grad_norm_([p], clip)
+
+            optimizer.step()
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:153]: It is actually significantly faster for this setup and uses negligible GPU memory compared to a single optimizer.step().

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                orig_dtype = p.data.dtype
+                p.data.copy_(
+                    (p.data.float() - _noise(p.data, n, self.current_seed).float() * self.current_std).to(orig_dtype)
+                )
+        self.current_seed, self.current_std = None, 0.0
+        return True
+
+    def apply_mutation(self, seed: Optional[int] = None, std: float = 0.0) -> Optional[int]:
+        if isinstance(seed, int) and std:
+            for n, p in self.model_runner.model.named_parameters():
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:56]: Removed those variables.

File: openrlhf/trainer/ray/vllm_engine.py [hunk 7]
```diff
@@ -145,6 +145,11 @@ async def wake_up(self, tags=["weights", "kv_cache"]):
         for tag in tags:
             await self.llm.wake_up(tags=[tag])
 
+    async def llm_generate_batch(self, prompts, sampling_params):
+        # AsyncLLMEngine.generate returns an async iterator; nothing runs until iterated.
+        # Iterating here makes the engine start immediately, without blocking the driver thread.
+        return [out async for out in self.llm.generate(prompts, sampling_params, request_id=random_uuid())]
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/vllm_engine.py]: Depreciated, so I removed llm_generate_batch.

File: examples/python/es_countdown_heuristic.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                    continue
+
+                clean_expr = re.sub(r"[^0-9+\-*/(). ]", "", raw_ans).strip()
+
+                if not clean_expr:
+                    rewards.append(0.0)
+                    continue
+
+                # 2. Use an empty dict {} for builtins instead of None to prevent TypeError
+                result = float(eval(clean_expr, {"__builtins__": {}}, {}))
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply examples/python/es_countdown_heuristic.py:54]: This is just an example that one can choose to run and is not essential to library functioning, so there is no security risk.

File: openrlhf/trainer/ray/es_worker_wrap.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py]: It is reversible. Do not make further edits regarding the float32 stuff; it was very carefully designed.

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+                for k, v in list(st.items()):
+                    if isinstance(v, torch.Tensor):
+                        st[k] = v.to(p.device)
+
+            # Compute ES gradient: weighted sum of noise vectors
+            g = torch.zeros_like(p, dtype=torch.float32)
+            for seed, w, _ in updates:
+                if isinstance(seed, int) and seed != STABILIZE_SEED:
+                    g.add_(_noise(p.data, name, seed).float(), alpha=float(w))
+            g.div_(max(1, len(updates)))
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:144]: noise returns a mean-zero, std-1 Gaussian tensor, so we do not need to divide by sigma here.

File: openrlhf/trainer/es_utils/generator.py [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+    ) -> List[ESExperience]:
+        sampling_params = SamplingParams(
+            temperature=generate_kwargs.get("temperature", 1.0),
+            top_p=generate_kwargs.get("top_p", 1.0),
+            top_k=generate_kwargs.get("top_k", -1),
+            max_tokens=generate_kwargs.get("max_new_tokens", 1024),
+            min_tokens=generate_kwargs.get("min_new_tokens", 1),
+            skip_special_tokens=generate_kwargs.get("skip_special_tokens", False),
+            logprobs=None,
+            stop=[generate_kwargs.get("stop_token", "</answer>")],
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/es_utils/generator.py]: Good suggestion I will make this change.

**Comment by DavidKoplow** [comment]
@gemini-code-assist re-run the review

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+
+def _noise(p: torch.Tensor, name: str, seed: int) -> torch.Tensor:
+
+    g = torch.Generator(device=p.device)
+    h = int.from_bytes(hashlib.sha256(name.encode()).digest()[:4], "little")
+    g.manual_seed((seed ^ h) & 0xFFFFFFFF)
+    n = torch.normal(0, 1, size=p.shape, device=p.device, generator=g, dtype=p.dtype)
+
+    return n
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:32]: It takes a few microseconds; this should not be cached.

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+            optimizer.step()
+            p.grad = None
+
+            # Move optimizer state back to CPU to save memory
+            st = optimizer.state.get(p)
+            if st:
+                for k, v in list(st.items()):
+                    if isinstance(v, torch.Tensor):
+                        st[k] = v.to("cpu")
+
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:162]: Memory doesn't allow, and the current implementation is insanely fast.

File: openrlhf/cli/serve_rm_v2.py [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+                    sanitized_extra_logs[key] = vals + [0.0] * (n - len(vals))
+                elif len(vals) > n:
+                    sanitized_extra_logs[key] = vals[:n]
+            for key in expected_log_keys:
+                if key not in sanitized_extra_logs:
+                    sanitized_extra_logs[key] = [0.0] * n
+
+            return JSONResponse({"rewards": rewards, "scores": rewards, "extra_logs": sanitized_extra_logs})
+        except Exception as e:
+            return JSONResponse({"error": str(e), "rewards": [], "scores": [], "extra_logs": {}}, status_code=500)
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/cli/serve_rm_v2.py:174]: Fixed

File: openrlhf/trainer/ray/es_worker_wrap.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+        # Revert any current mutation before applying gradient
+        self.revert_mutation()
+
+        if len(updates) > 1:
+            sc = torch.tensor([x[1] for x in updates], dtype=torch.float32)
+            sd = sc.std()
+            sc = (sc - sc.mean()) / sd if sd > 1e-8 else (sc - sc.mean())
+            updates = [(seed, float(v), sig) for (seed, _, sig), v in zip(updates, sc)]
+
+        clip = float(os.getenv("ES_CLIP_GRAD_NORM", "0.0"))
```
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:126]: Fixed

**Comment by DavidKoplow** [comment]
@gemini-code-assist re-run the review

File: openrlhf/trainer/ray/es_worker_wrap.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:144]: It doesn't need this since all sigmas are the same, and w is already normalized and _noise has std = 1.

**Comment by DavidKoplow** [comment]
@gemini-code-assist re-run the review

File: openrlhf/trainer/ray/es_worker_wrap.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:32]: Manual seed doesn't work because it has issues with vLLM.

File: openrlhf/trainer/ray/es_worker_wrap.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DavidKoplow) [reply openrlhf/trainer/ray/es_worker_wrap.py:161]: This is fine and seems very stable and fast.

**Comment by DavidKoplow** [comment]
@gemini-code-assist re-run the review

**Comment by DavidKoplow** [comment]
@gemini-code-assist, re-run the review and keep a careful track of code comments before saying something is an issue.
